### PR TITLE
Refactor to use fetch helper

### PR DESCRIPTION
**Pull request by francislata**
This PR refactors the codebase to use the new `fetch` helper function introduced in #2363.

~In addition, I have added the ability to specify an extension when using the `fetch()` helper function.~

Also, I have removed some of the download fetching helper functions in `extra/utils.py` as well as the associated test suite.

### CODE REVIEW

File: examples/whisper.py [hunk 1]
```diff
@@ -249,9 +245,7 @@ def listener(q):
 def init_whisper(model_name="tiny.en", batch_size=1):
   assert MODEL_URLS[model_name] is not None
 
-  filename = BASE / "whisper-{}.pt".format(model_name)
```
Comment (mmmkkaaayy) [inline examples/whisper.py:252]: The `whisper-` prefix was used during CI caching

Comment (francislata) [reply examples/whisper.py:252]: yeah good call. i'm definitely going to go through all of these again just to make sure I don't break any existing caching, etc.

Comment (francislata) [reply examples/whisper.py:252]: I updated the `workflow.yml` file to look at the new cache paths now.

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -85,8 +85,8 @@ jobs:
       uses: actions/cache@v3
       with:
         path: |
-          weights/whisper-tiny.en.pt
-          weights/gpt2.tiktoken
+          ~/.cache/tinygrad/downloads/3a2cae277fcec4464349df535729caaf
```
Comment (geohot) [inline .github/workflows/test.yml:88]: I don't like this with the hashes, it's hard to read. The whole downloads folder should probably be cached

File: tinygrad/helpers.py [hunk 3]
```diff
@@ -36,6 +38,17 @@ def partition(lst:List[T], fxn:Callable[[T],bool]):
   b:List[T] = []
   for s in lst: (a if fxn(s) else b).append(s)
   return a,b
+def temp(x:str) -> str: return (Path(tempfile.gettempdir()) / x).as_posix()
+def get_child(parent, key):
```
Comment (geohot) [inline tinygrad/helpers.py:42]: Can this be a separate PR?

File: tinygrad/helpers.py [hunk 4]
```diff
@@ -223,8 +236,9 @@ def wrapper(*args, **kwargs) -> bytes:
 
 # *** http support ***
 
-def fetch(url:str) -> pathlib.Path:
+def fetch(url:str, ext:Optional[str]=None) -> pathlib.Path:
```
Comment (geohot) [inline tinygrad/helpers.py:239]: Why do I need an extension?

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -85,8 +85,8 @@ jobs:
...
+          ~/.cache/tinygrad/downloads/5386a662f5b1245493601b29a5e61f40
         key: model-weights-v1
```
Comment (mmmkkaaayy) [inline .github/workflows/test.yml:90]: Should update the key as well

File: tinygrad/helpers.py (hunk 4 above)
Comment (francislata) [reply tinygrad/helpers.py:239]: The `test/external/external_test_yolov8.py` was giving issues complaining that there's no ".pt" extension. I'll update the test to just rename the cache directly to have the extension.

File: tinygrad/helpers.py (hunk 3 above)
Comment (francislata) [reply tinygrad/helpers.py]: Will do this in a separate PR. I'll link the PR number to this comment.

File: .github/workflows/test.yml [hunk 6]
```diff
@@ -85,9 +85,8 @@ jobs:
       uses: actions/cache@v3
       with:
         path: |
```
Comment (geohot) [inline .github/workflows/test.yml:87]: Clean this up. Why does it have the pipe? It is only caching model weights?

File: examples/compile_efficientnet.py [hunk 7]
```diff
@@ -21,10 +20,12 @@
   else:
     cprog = [prg]
     # image library!
-    cprog += ["#define STB_IMAGE_IMPLEMENTATION", fetch("https://raw.githubusercontent.com/nothings/stb/master/stb_image.h").decode('utf-8').replace("half", "_half")]
+    with open(fetch("https://raw.githubusercontent.com/nothings/stb/master/stb_image.h"), "rb") as f:
```
Comment (geohot) [inline examples/compile_efficientnet.py:23]: pathlib has a read method afaik

File: examples/efficientnet.py [hunk 8]
```diff
@@ -61,7 +60,8 @@ def infer(model, img):
   model.load_from_pretrained()
 
   # category labels
-  lbls = fetch("https://gist.githubusercontent.com/yrevar/942d3a0ac09ec9e5eb3a/raw/238f720ff059c1f82f368259d1ca4ffa5dd8f9f5/imagenet1000_clsidx_to_labels.txt")
+  with open(fetch("https://gist.githubusercontent.com/yrevar/942d3a0ac09ec9e5eb3a/raw/238f720ff059c1f82f368259d1ca4ffa5dd8f9f5/imagenet1000_clsidx_to_labels.txt"), "rb") as f:
```
Comment (geohot) [inline examples/efficientnet.py:63]: same as above, this is more unneeded lines

File: tinygrad/helpers.py [hunk 9]
```diff
@@ -13,6 +13,7 @@ def prod(x:Iterable[T]) -> Union[T,int]: return functools.reduce(operator.__mul_
 
 # NOTE: helpers is not allowed to import from anything else in tinygrad
 OSX = platform.system() == "Darwin"
+WINDOWS = platform.system() == "Windows"
```
Comment (geohot) [inline tinygrad/helpers.py:16]: Why is this here?

**Comment by geohot** [comment]
This PR is way too large to review. Want to start with only the whisper cache?

**Comment by francislata** [comment]
> This PR is way too large to review. Want to start with only the whisper cache?

Yeah, sounds good. I'll separate these out into smaller PRs.
